src/wui/inputqueuedisplay.cc

```diff
@@ -19,586 +19,574 @@
+               kButtonSize * 3 / 2,
+               UI::ButtonStyle::kWuiMenu,
+               "",
+               "",
+               UI::Button::VisualState::kFlat),
+     priority_(&hbox_,
+               0,
+               0,
+               5 * kButtonSize,
+               kButtonSize,
+               0,
+               4,
+               has_priority_ ?
+                  priority_to_index(settings_ ? settings_->ware_queues.at(index_).priority :
+                                                building_.get_priority(type_, index_)) :
+                  2,
+               UI::SliderStyle::kWuiLight,
+               "",
+               kButtonSize,
+               can_act_ && has_priority_),
+     spacer_(&hbox_, 0, 0, priority_.get_w(), priority_.get_h()),
+     slider_was_moved_(nullptr),
+     collapsed_(false),
+     nr_icons_(queue_ ?
+                  queue_->get_max_size() :
+                  type_ == Widelands::wwWORKER ? settings_->worker_queues.at(index_).max_fill :
+                                                 settings_->ware_queues.at(index_).max_fill),
+     icons_(nr_icons_, nullptr),
+     fill_index_under_mouse_(-1) {
+
+	assert((queue_ == nullptr) ^ (settings_ == nullptr));
+
+	b_decrease_real_fill_.set_visible(queue_ && ibase_.omnipotent());
+	b_increase_real_fill_.set_visible(queue_ && ibase_.omnipotent());
+
+	b_decrease_desired_fill_.set_visible(!show_only_);
+	b_increase_desired_fill_.set_visible(!show_only_);
+	b_decrease_desired_fill_.set_enabled(can_act_);
+	b_increase_desired_fill_.set_enabled(can_act_);
+
+	b_increase_desired_fill_.set_repeating(true);
+	b_increase_real_fill_.set_repeating(true);
+	b_decrease_desired_fill_.set_repeating(true);
+	b_decrease_real_fill_.set_repeating(true);
+
+	vbox_.add_space(kButtonSize / 4);
+
+	hbox_.add(&b_decrease_real_fill_);
+	hbox_.add(&b_decrease_desired_fill_);
+
+	for (size_t i = 0; i < nr_icons_; ++i) {
+		icons_[i] = new UI::Icon(&hbox_, 0, 0, kButtonSize, kButtonSize,
+		                         type_ == Widelands::wwWARE ?
+		                            building_.owner().tribe().get_ware_descr(index_)->icon() :
+		                            building_.owner().tribe().get_worker_descr(index_)->icon());
+		hbox_.add(icons_[i]);
+	}
+
+	hbox_.add_inf_space();
+	hbox_.add(&b_increase_desired_fill_);
+	hbox_.add(&b_increase_real_fill_);
+
+	priority_.set_cursor_fixed_height(kButtonSize * 2 / 3);
+
+	// To make sure the fill buttons are aligned even when some queues
+	// have priority buttons and some don't (e.g. in barracks)
+	hbox_.add_space(kButtonSize / 4);
+	priority_.set_visible(has_priority_);
+	spacer_.set_visible(!has_priority_);
+	hbox_.add(&priority_, UI::Box::Resizing::kAlign, UI::Align::kCenter);
+	hbox_.add(&spacer_);
+
+	vbox_.add(&hbox_, UI::Box::Resizing::kFullSize);
+	vbox_.add_space(kButtonSize / 4);
+	add(&vbox_, UI::Box::Resizing::kExpandBoth);
+
+	add_space(kButtonSize / 4);
+	add(&collapse_);
+
+	collapse_.sigclicked.connect([this]() {
```

Comment:
How about always collapsing when `!can_act_`, and not allowing to expand? This would make the UI more compact for spectators.